workflow/controller/operator_concurrency_test.go

```diff
@@ -1046,3 +1048,95 @@ func TestSynchronizationForPendingShuttingdownWfs(t *testing.T) {
 		assert.Equal(t, wfv1.WorkflowRunning, wocTwo.execWf.Status.Phase)
 	})
 }
+
+func TestWorkflowMemoizationWithMutex(t *testing.T) {
+	wf := wfv1.MustUnmarshalWorkflow(`apiVersion: argoproj.io/v1alpha1
+kind: Workflow
+metadata:
+  generateName: example-steps-simple
+  namespace: default
+spec:
+  entrypoint: main
+  templates:
+    - name: main
+      steps:
+        - - name: job-1
+            template: sleep
+            arguments:
+              parameters:
+                - name: sleep_duration
+                  value: 10
+          - name: job-2
+            template: sleep
+            arguments:
+              parameters:
+                - name: sleep_duration
+                  value: 5
+
+    - name: sleep
+      synchronization:
+        mutex:
+          name: mutex-example-steps-simple
+      inputs:
+        parameters:
+          - name: sleep_duration
+      script:
+        image: alpine:latest
+        command: [/bin/sh]
+        source: |
+          echo "Sleeping for {{ inputs.parameters.sleep_duration }}"
+          sleep {{ inputs.parameters.sleep_duration }}
+      memoize:
+        key: "memo-key-1"
+        cache:
+          configMap:
+            name: cache-example-steps-simple
+    `)
+	cancel, controller := newController(wf)
+	defer cancel()
+
+	ctx := context.Background()
+
+	woc := newWorkflowOperationCtx(wf, controller)
+	woc.operate(ctx)
+
+	holdingJobs := make(map[string]string)
+	for _, node := range woc.wf.Status.Nodes {
+		holdingJobs[node.ID] = node.DisplayName
+	}
+
+	// Check initial status: job-1 acquired the lock
+	job1AcquiredLock := false
+	if woc.wf.Status.Synchronization != nil && woc.wf.Status.Synchronization.Mutex != nil {
+		for _, holding := range woc.wf.Status.Synchronization.Mutex.Holding {
+			if holdingJobs[holding.Holder] == "job-1" {
+				fmt.Println("acquired: ", holding.Holder)
+				job1AcquiredLock = true
+			}
+		}
+	}
+	assert.True(t, job1AcquiredLock)
+
+	// Make job-1's pod succeed
+	makePodsPhase(ctx, woc, apiv1.PodSucceeded, func(pod *apiv1.Pod) {
+		if pod.ObjectMeta.Name == "job-1" {
+			pod.Status.Phase = apiv1.PodSucceeded
+		}
+	})
+	woc.operate(ctx)
+
+	// Check final status: both job-1 and job-2 succeeded, job-2 simply hit the cache
+	job1CacheHit, job2CacheHit := false, false
+	for _, node := range woc.wf.Status.Nodes {
+		switch node.DisplayName {
+		case "job-1":
+			assert.Equal(t, wfv1.NodeSucceeded, node.Phase)
+			job1CacheHit = node.MemoizationStatus.Hit
+		case "job-2":
+			assert.Equal(t, wfv1.NodeSucceeded, node.Phase)
+			job2CacheHit = node.MemoizationStatus.Hit
+		}
+	}
+	assert.False(t, job1CacheHit)
```

Comment:
nit: Why not just assert in the case above (line 1134), it feels more readble.

Reply:
Nice catch. It seems really weird... why did I do that? :)